sql-to-dbsp-compiler/lib/sqllib/src/operators.rs

```diff
@@ -86,22 +86,144 @@ for_all_compare!(gte, bool);
 #[inline(always)]
 fn plus<T>(left: T, right: T) -> T
 where
-    T: Add<Output = T>,
+    T: CheckedAdd,
 {
+    left.checked_add(&right)
+        .expect("attempt to add with overflow")
+}
+
+for_all_int_operator!(plus);
+
+pub fn plus_f_f(left: F32, right: F32) -> F32 {
+    left + right
+}
+
+pub fn plus_fN_f(left: Option<F32>, right: F32) -> Option<F32> {
+    let left = left?;
+    Some(plus_f_f(left, right))
+}
+
+pub fn plus_f_fN(left: F32, right: Option<F32>) -> Option<F32> {
+    let right = right?;
+    Some(plus_f_f(left, right))
+}
+
+pub fn plus_fN_fN(left: Option<F32>, right: Option<F32>) -> Option<F32> {
+    let left = left?;
+    let right = right?;
+    Some(plus_f_f(left, right))
+}
+
+pub fn plus_d_d(left: F64, right: F64) -> F64 {
     left + right
 }
 
-for_all_numeric_operator!(plus);
+pub fn plus_dN_d(left: Option<F64>, right: F64) -> Option<F64> {
+    let left = left?;
+    Some(plus_d_d(left, right))
+}
+
+pub fn plus_d_dN(left: F64, right: Option<F64>) -> Option<F64> {
+    let right = right?;
+    Some(plus_d_d(left, right))
+}
+
+pub fn plus_dN_dN(left: Option<F64>, right: Option<F64>) -> Option<F64> {
+    let left = left?;
+    let right = right?;
+    Some(plus_d_d(left, right))
+}
+
+pub fn plus_decimal_decimal(left: Decimal, right: Decimal) -> Decimal {
```

Comment:
I think decimal addition can also overflow.
we are not checking here the type, but what happens if you exceed the 128 bits of precision of rust_decimal?
does addition panic?

Reply:
Actually yeah, decimal also panics. Will use checked operation there too.